zerver/views/auth.py

```diff
@@ -453,6 +454,32 @@ def start_social_signup(request: HttpRequest, backend: str, extra_arg: Optional[
     return oauth_redirect_to_root(request, backend_url, 'social', is_signup=True,
                                   extra_url_params=extra_url_params)
 
+def social_auth(request: HttpRequest, backend: str) -> HttpResponse:
+    """
+    python-social-auth sets certain fields from the request into the session
+    and doesn't clear them if another request is made with a field that was present
+    in the previous request now missing. We use this function to hook into the beginning
+    of the social auth flow to ensure the session is properly cleared out.
+    This function and the corresponding url entry in urls.py should be removed if this issue
+    gets fixed upstream - https://github.com/python-social-auth/social-core/issues/425
+    """
+
+    # This block replicates how python-social-auth chooses whether to take params
+    # from POST or GET.
+    if request.method == 'POST':  # nocoverage
+        data = request.POST
+    else:
+        data = request.GET
+
+    for field_name in settings.SOCIAL_AUTH_FIELDS_STORED_IN_SESSION:
+        if field_name not in data:
```

Comment:
Do we need this `if` statement?  I'd think we could just delete every value from the session here.